[MM-44655] : Add channel files screen

AshishDhama opened the pull request:
#### Summary
Add channel files screen
Note: Global file search should also be tested when testing this as we have done some refactoring.

#### Ticket Link
  Fixes https://mattermost.atlassian.net/browse/MM-44655

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: <!-- Device name(s), OS version(s) -->

#### Screenshots

https://user-images.githubusercontent.com/16203333/229535433-f514f3d0-f577-41a3-91af-5fb0cd62dfcf.mov

https://user-images.githubusercontent.com/16203333/229537278-ac8014db-26b5-4d24-9d77-10776af1278a.mov]

       
| Images | |
|--------|--------|
| ![Screenshot_1680773479](https://user-images.githubusercontent.com/16203333/230336524-c6513f29-596e-44ed-b047-1295135eb062.png) | ![Screenshot_1680773460](https://user-images.githubusercontent.com/16203333/230337660-7940f6d7-d7e3-4e99-9d71-7d664e68c685.png) |
| ![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-04-06 at 15 05 56](https://user-images.githubusercontent.com/16203333/230337527-a8a90aca-92d4-40cf-a032-4a5d4decfb01.png) | ![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-04-06 at 15 05 46](https://user-images.githubusercontent.com/16203333/230337596-11ff46bc-472f-4719-9038-46c03af589b0.png)| 
| ![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-04-06 at 15 00 45](https://user-images.githubusercontent.com/16203333/230337586-f9939443-0b8e-459e-b251-41a4efc24bff.png)| |



#### Release Note
<!--
Add a release note for each of the following conditions:

* New features and improvements, including behavioural changes, UI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Example:

```release-note
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->

```release-note
- Add channel files screen
```

larkox commented:
@itao @matthewbirtch This PR adds the file count to the channel info, and also the files screen.

The file count was added on 6.7, if I remember correctly. The file search (used to get all the files in the channel) on 5.34.

I think you know better the state of the market, so... should @AshishDhama do anything special for versions prior to 6.7 (or even prior to 5.34)?

If he does nothing, on those servers, no matter how many files there are, it will show the Files option with a 0, as in "there are no files in this channel" no matter if there are files or not.

Also, for servers prior to 5.34, going into the files screen, in the best scenario, will show as if there are no files.

Keep in mind that the app opens as long the server version is above 5.26.2.

matthewbirtch commented:
If it's trivial to do a server version check to hide the file count for servers that don't support it, I'd be all for that. But, if it's more involved, based on what @enahum said, it's likely not worth it since there are other potential issues on unsupported server versions

AshishDhama commented:
@enahum @larkox Right now I am testing it, But I think its ready for first round of reviews.

enahum commented on app/screens/channel_files/channel_files.tsx:
```diff
@@ -0,0 +1,212 @@
+import {useTheme} from '@context/theme';
+import {useIsTablet} from '@hooks/device';
+import {useImageAttachments} from '@hooks/files';
+import {showMobileOptionsBottomSheet} from '@screens/home/search/results/file_options/mobile_options';
+import Toasts from '@screens/home/search/results/file_options/toasts';
+import FileResult from '@screens/home/search/results/file_result';
```
I think these should be moved from `@screens/home/search`  to some utility file an components in common

AshishDhama replied:
What should be the ideal place to keep common code.

enahum commented on app/screens/channel_files/channel_files.tsx:
this file at first glance, looks very similar to the functionality to display files in the search screen, I wonder if most functionality can be combined in the same file

AshishDhama replied:
It is not exactly same, and the common components are being reused.

enahum commented on app/screens/channel_files/empty/index.tsx:
```diff
@@ -0,0 +1,56 @@
+    },
+}));
+
+function EmptySavedPosts() {
```
I think this was a copy/paste from another component based on the name of the component, what is different between the two components? can they be combined into one and handle the diff with props?

AshishDhama replied:
Yeah I forgot to change the name, Have done the required refactoring.

enahum commented:
@AshishDhama added some quick feedback in order to avoid duplicating code as much as possible, let me know if you want me to review more in depth

larkox commented on app/screens/channel_files/channel_files.tsx:
This file seems to have been extracted from `@screens/home/search/results/file_results.tsx`. I check, and some of the comments I made here are also present on that file, so it would be great if you could address them also in the other file (or as stated before, try to somehow mix as much as possible in a common component).

larkox commented on app/screens/channel_files/header.tsx:
```diff
@@ -0,0 +1,126 @@
+    const isTablet = useIsTablet();
+    const hasFilters = selectedFilter !== FileFilters.ALL;
+
+    let messagesText = intl.formatMessage({id: `screen.${TEST_ID}.header.recent_files`, defaultMessage: 'Recent Files'});
```
Does this behave correctly with `i18n-extract`? I would be happily surprised if it did, but please double check.

larkox commented on app/screens/channel_files/index.ts:
```diff
@@ -0,0 +1,32 @@
+
+const enhance = withObservables(['channelId'], ({channelId, database}: Props) => {
+    const team = observeCurrentTeam(database);
+    const channel = observeChannel(database, channelId);
```
Why do you observe the channel if you only want the id, and the id comes from the props?

larkox commented on app/screens/channel_files/index.ts:
```diff
@@ -0,0 +1,32 @@
+    const team = observeCurrentTeam(database);
+    const channel = observeChannel(database, channelId);
+    return {
+        teamId: team.pipe(switchMap((t) => of$(t?.id))),
```
Do we want always the current team id? Or do you foresee any case where we may be looking at a channel in a different team? If that is the case, it would make sense to get the channel team id, and default to the current team id (for DM/GMs).

larkox reviewed (changes requested):
A few comments.

mattermost-build commented:
Building app in separate branch.

AshishDhama commented:
@larkox except these I think rest all are done.
<img width="863" alt="Screenshot 2023-04-05 at 3 27 15 PM" src="https://user-images.githubusercontent.com/16203333/230047808-86171cd4-6a83-4d2a-8ea2-f45402075cac.png">
Testing it.

<img width="863" alt="Screenshot 2023-04-05 at 3 26 41 PM" src="https://user-images.githubusercontent.com/16203333/230047787-cdadd2e1-406a-446e-af40-f7ef9b7f715e.png">
Not sure if should remove this, I fear breaking some edge case.

larkox commented on app/components/files_search/file_results.tsx:
The useMemos around `orderedFileInfos` seem off. The dependencies don't seem to match what they should. Maybe the I am missing some information having to do with `filesForGallery` (for example) not changing any time, but something feels off. Do you mind double checking?

enahum replied:
this seems resolved.

larkox commented on app/screens/channel_files/channel_files.tsx:
```diff
@@ -0,0 +1,165 @@
+        } else {
+            handleSearch(term, filter);
+        }
+    }, [filter, term]);
```
This may depend on `handleSearch`. 0/5 on whether we should trigger the effect if handleSearch changes (probably we do, since that would mean the channel or the server changed)

larkox commented on app/screens/channel_files/channel_files.tsx:
```diff
@@ -0,0 +1,256 @@
+            />
+        );
+    }, [
+        (orderedFileInfos.length === 1) && orderedFileInfos[0].mime_type,
```
Let's keep it for now. we can revisit it later.

larkox reviewed (approved):
LGTM, just small things to check.

matthewbirtch reviewed (changes requested):
Aweseom work @AshishDhama. This is working great. Just two things from me:

1. Since this is a files view for a specific channel, we don't need to show the channel name label that shows within the file attachment card since it will all be the same channel name (and the channel name is shown in the top bar)
2. When filtering is active, the dot that displays on the filter icon is positioned too low. Can you bump it up - looks like it needs to move up by 6 or 8 (note: this may also apply to the global file search filter, but would outside of this scope)
    <img width="564" alt="Screenshot 2023-04-05 at 2 24 01 PM" src="https://user-images.githubusercontent.com/2040554/230172584-d57238b8-8ec1-4f10-bf98-5c9d4ebbb811.png">
3. The file results load as I'm typing, which is fine, but the results are a little confusing for me. If I type `i` it shows no results. If I type `im` it also shows no results, but if I type `imag` it will show results with filenames that have `image` in the filename. Is this expected?

Other than that, I don't have much else that I've come across yet. I have yet to test on iPad or android, but wanted to share these pieces of feedback

AshishDhama commented:
> 3. The file results load as I'm typing, which is fine, but the results are a little confusing for me. If I type `i` it shows no results. If I type `im` it also shows no results, but if I type `imag` it will show results with filenames that have `image` in the filename. Is this expected?

@matthewbirtch Yeah, it's the way search works right now, Harshil from core 1 is working on to change this.

AshishDhama commented:
/update-branch

AshishDhama commented:
@matthewbirtch I have done all the required changes, Fixed the dot position in global search as well.
@larkox Yeah, Bottom menu is still valid, I have added screenshots in the description.

matthewbirtch reviewed (changes requested):
Looks good @AshishDhama. Two other minor details I noticed this round:

- The search bar doesn't seem to be spanning the same width and is showing extra margin on the right side.
<img width="564" alt="Screenshot 2023-04-06 at 11 43 37 AM" src="https://user-images.githubusercontent.com/2040554/230431637-cb47884f-0f1d-4202-9021-92911fd35372.png">
- When I landing on the channel files view, If I focus on the search field and then cancel without typing anything, the spinner displays infinitely for me. It shouldn't even try to load anything if I have not typed anything and just cancel.

<img src="https://user-images.githubusercontent.com/2040554/230433152-5c2b9194-318b-437f-8aa4-f85d98867485.gif" width="450">

larkox commented on app/screens/channel_files/channel_files.tsx:
```diff
@@ -0,0 +1,167 @@
+        setLoading(false);
+    }, [serverUrl, channel]);
+
+    useEffect(() => {
+        if (term) {
+            if (searchTimeoutId.current) {
+                clearTimeout(searchTimeoutId.current);
+            }
+
+            searchTimeoutId.current = setTimeout(() => {
+                handleSearch(term, filter);
+            }, General.SEARCH_TIMEOUT_MILLISECONDS);
+        } else {
+            handleSearch(term, filter);
+        }
+    }, [filter, term, handleSearch]);
```
This is going to be fun :smile:

Image you write something, and fast you delete it. For example, `a`.
The moment you write `a`, this effect triggers. Since you have  a term (`a`), it will go through the first part of the if, setting a timeout that will eventually execute search (lets call it `search a`).
But then we delete `a`, so this effect will trigger again, and will go through the else branch, calling `handleSearch` directly (lets call it `search all`).

`search all` will execute first, and if everything happens as expected on the network, it will return first, and will set the file infos and loading to the correct values.
But the timeout for `search a` is still running, and will execute, creating a new network call, and eventually, updating again the file infos.

So the user experience will be:
- I write `a`
- I delete the `a`
- I see the loading spinner
- I see the correct results
- Suddenly, the correct results transform into a subset of the results

The simplest solution is to remove the if/else at all. You are doing practically the same on both branches.

That will still give some problems if the network calls return in a different order, but that is way less common in this scenario. Just for fun, let me show you the needed steps for that to happen:
- I write `a`
- I wait for `SEARCH_TIMEOUT_MILLISECONDS` so `handleSearch` is executed (`search a`)
- I write `ab`
- I wait for `SEARCH_TIMEOUT_MILLISECONDS` so `handleSearch` is executed (`search ab`)
- `search a` for whatever reason is taking a lot of time
- `search ab`, on the other hand, returns faster, and sets the file infos
- `search a` finally returns, and set also the file infos.
- The user experience becomes "I had the right results, and now I have more results than I should have"

If we were to address this, the best way would be to store the `searchFiles` request in a reference, and cancel it, but we still don't have support for that. The next best way to address that is to store a timestamp like `lastSearchRequest`. That way, handle search can do something like:
```
const t = Date.now()
lastSearchRequest.current = t;
const {files} = await searchFiles(...);
if (lastSearchRequest.current !== t) {
    return;
}
...
```

larkox commented on app/screens/channel_files/channel_files.tsx:
```diff
@@ -0,0 +1,167 @@
+                />
+            }
+            {!loading &&
+                <>
```
This fragment seems unneeded.

larkox reviewed (changes requested):
Some minor changes.

matthewbirtch reviewed (dismissed):
This is looking/working great for me @AshishDhama. Nice work!

larkox reviewed (approved):
LGTM, great work!

matthewbirtch commented:
@AshishDhama can we make one minor change I just noticed?

When filtering files, if I get no results with the applied filter, the empty state says: 

**No files yet**
Files posted in this channel will show here.

When we get empty results from a filter, can we change it to:

**No files found**
This channel doesn't contain any files with the applied filters.

enahum commented on app/components/files_search/file_results.tsx:
```diff
@@ -88,10 +91,14 @@ const FileResults = ({
     const renderItem = useCallback(({item}: ListRenderItemInfo<FileInfo>) => {
+        let channelName: string | undefined;
+        if (!isChannelFiles) {
+            channelName = channelNames[item.channel_id!];
```
```suggestion
            channelName = channelNames[item.channel_id];
```

enahum commented on app/components/files_search/file_results.tsx:
```diff
@@ -88,10 +93,14 @@ const FileResults = ({
 
     const renderItem = useCallback(({item}: ListRenderItemInfo<FileInfo>) => {
+        let channelName: string | undefined;
+        if (!isChannelFiles) {
```
```suggestion
        if (!isChannelFiles && item.channel_id) {
```

enahum commented on app/screens/channel_files/channel_files.tsx:
```diff
@@ -0,0 +1,169 @@
+        setTerm('');
+    }, []);
+
+    const onTextChange = useCallback((searchTerm: string) => {
```
I think we want to debounce this so that is not performed on every key press, it looks like too much at the moment, also there is no request sequence control, so if one request finishes before the other, results may not match the search

enahum replied:
nvm I see that you have a use effect with a timeout for this

enahum commented on app/screens/channel_files/channel_files.tsx:
```diff
@@ -0,0 +1,169 @@
+        lastSearchRequest.current = t;
+        const searchParams = getSearchParams(channel.id, searchTerm, ftr);
+        const {files} = await searchFiles(serverUrl, channel.teamId, searchParams);
+        if (lastSearchRequest.current !== t) {
+            return;
+        }
```
I'm failing to see when this assertion is true, is there really a need for `lastSearchRequest` ?

enahum commented on app/screens/channel_files/header.tsx:
```diff
@@ -0,0 +1,126 @@
+    const isTablet = useIsTablet();
+    const hasFilters = selectedFilter !== FileFilters.ALL;
+
+    let messagesText = intl.formatMessage({id: 'screen.channel_files.header.recent_files', defaultMessage: 'Recent Files'});
+    if (hasFilters) {
+        messagesText = intl.formatMessage({id: FilterData[selectedFilter].id, defaultMessage: FilterData[selectedFilter].defaultMessage});
+    }
```
the check can be done differently to avoid one computation here.

```suggestion
    let messagesText = hasFilters : intl.formatMessage({id: FilterData[selectedFilter].id, defaultMessage: FilterData[selectedFilter].defaultMessage}); ? intl.formatMessage({id: 'screen.channel_files.header.recent_files', defaultMessage: 'Recent Files'});
```

the above suggestion properly formatted.

enahum commented on app/screens/channel_files/channel_files.tsx:
```diff
@@ -0,0 +1,169 @@
+            {!loading &&
+            <FileResults
+                canDownloadFiles={canDownloadFiles}
+                fileChannels={[channel]}
```
you should memoize the array

enahum commented on app/screens/channel_files/channel_files.tsx:
```diff
@@ -0,0 +1,169 @@
+                <Loading
+                    color={theme.buttonBg}
+                    size='large'
+                    containerStyle={[styles.loading]}
```
```suggestion
                    containerStyle={styles.loading}
```

enahum commented on app/screens/channel_files/channel_files.tsx:
```diff
@@ -0,0 +1,169 @@
+                canDownloadFiles={canDownloadFiles}
+                fileChannels={[channel]}
+                fileInfos={fileInfos}
+                paddingTop={{paddingTop: 0}}
```
you should define a style for this and use the style here instead of inline

enahum reviewed (changes requested):
great work @AshishDhama just a few small details and we can merge this.
